### PR TITLE
[dv/dvsim] Group failures per test in buckets

### PR DESCRIPTION
**Pull request by matutem**
Within each bucket collect failures per unqualified test name.
Cap the failures shown to 5 per unqualified test name.
Show the number of failures per bucket and per test.
Use a trailing backslash to separate the line and log info from the
qualified test name.

Signed-off-by: Guillermo Maturana <maturana@google.com>

### CODE REVIEW

**Review by udinator — APPROVED** [review]
LGTM, I think this makes sense overall to me.

File: util/dvsim/SimCfg.py [hunk 1]
```diff
@@ -539,20 +542,65 @@ def _gen_results(self, run_results):
...
+            by_tests = sorted(buckets.items(),
+                              key=lambda i: len(i[1]),
+                              reverse=True)
+            fail_msgs = ["\n## Failure Buckets", ""]
+            for bucket, tests in by_tests:
+                fail_msgs.append(f"* `{bucket}` has {len(tests)} failures:")
+                unqual_dict = collections.defaultdict(list)
+                for (test, line, context) in tests:
+                    unqual_dict[test.name].append((test, line, context))
+                for name, qual_tests in unqual_dict.items():
```
[the remark this answers is not in the document]
Comment (matutem) [reply util/dvsim/SimCfg.py]: Done.

File: util/dvsim/SimCfg.py [hunk 2]
```diff
@@ -539,20 +542,65 @@ def _gen_results(self, run_results):
         is enabled, then the summary coverage report is also generated. The final
         result is in markdown format.
         '''
+        def create_n_spaces(n):
```
[the remark this answers is not in the document]
Comment (matutem) [reply util/dvsim/SimCfg.py]: Sounds good. Note indenting the snippet of the log needs an additional two levels (so it gets treated as a code block).

**Review by sriyerg — APPROVED** [review]
LGTM, thanks!

File: util/dvsim/SimCfg.py [hunk 3]
```diff
@@ -539,20 +543,67 @@ def _gen_results(self, run_results):
...
+                unqual_dict = collections.defaultdict(list)
+                for (test, line, context) in tests:
+                    unqual_dict[test.name].append((test, line, context))
+                for test_count, (name, qual_tests) in enumerate(
+                        unqual_dict.items()):
+                    if test_count == _MAX_TESTS:
+                        fail_msgs.append(
+                            f"{indent_by(1)}* ... and "
+                            f"{len(unqual_dict) - test_count} more tests.")
+                        break
```
Comment (sriyerg) [inline util/dvsim/SimCfg.py:593]: Sorry, just read this part of code a bit more carefully. The `...more tests` computation seems wrong. Should it be: `sum(len(t) for t in unqual_dict) - running_count`, where `running_count` is the running total of tests that have been accounted for?

File: util/dvsim/SimCfg.py [hunk 4]
```diff
@@ -539,20 +543,67 @@ def _gen_results(self, run_results):
...
+            Returns:
+              A list of text lines for the report.
+            """
+            by_tests = sorted(buckets.items(),
+                              key=lambda i: len(i[1]),
+                              reverse=True)
+            fail_msgs = ["\n## Failure Buckets", ""]
+            for bucket, tests in by_tests:
+                fail_msgs.append(f"* `{bucket}` has {len(tests)} failures:")
+                unqual_dict = collections.defaultdict(list)
```
Comment (sriyerg) [inline util/dvsim/SimCfg.py:584]: minor nit: the qual v/s unqual variable naming made it a bit hard for me to follow the code. How about these:
unqual_dict -> unique_tests
qual_tests -> test_reseeds
_MAX_TESTS -> _MAX_UNIQUE_TESTS
_MAX_QUAL_TESTS_PER_TEST -> _MAX_TEST_RESEEDS

`SimCfg` already understands the concept of test reseeds.

File: util/dvsim/SimCfg.py [hunk 5]
```diff
@@ -539,20 +543,67 @@ def _gen_results(self, run_results):
...
+                for test_count, (name, qual_tests) in enumerate(
+                        unqual_dict.items()):
+                    if test_count == _MAX_TESTS:
+                        fail_msgs.append(
+                            f"{indent_by(1)}* ... and "
+                            f"{len(unqual_dict) - test_count} more tests.")
+                        break
+                    fail_msgs.append(f"{indent_by(1)}* Test {name} has "
+                                     f"{len(qual_tests)} failures.")
+                    for count, (test, line, context) in enumerate(qual_tests):
```
Comment (sriyerg) [inline util/dvsim/SimCfg.py:596]: The `qual_tests` is a list. I think we can probably just slice the array to limit what's printed. That way we can avoid the looping construct and having check each iteration. 

Come to think of it  - perhaps the earlier for loop could follow the same.

**Review by sriyerg — CHANGES_REQUESTED** [review]
Sorry, I am reverting my 'approval' because there seems to be a bug in the code. Let me know if you agree. Thanks!

File: util/dvsim/SimCfg.py (hunk 4 above)
Comment (matutem) [reply util/dvsim/SimCfg.py]: Done.

File: util/dvsim/SimCfg.py (hunk 5 above)
Comment (matutem) [reply util/dvsim/SimCfg.py]: More pythonic, thanks. Notice the top dict.items() needs to be cast to a list for slicing.

**Review by sriyerg — APPROVED** [review]
LGTM, thanks!

File: util/dvsim/SimCfg.py [hunk 6]
```diff
@@ -539,20 +543,67 @@ def _gen_results(self, run_results):
...
+                            create_failure_message(test, line, context))
+                    if len(test_reseeds) > _MAX_TEST_RESEEDS:
+                        fail_msgs.append(
+                            f"{indent_by(2)}* ... and "
+                            f"{len(test_reseeds) - _MAX_TEST_RESEEDS} "
+                            "more failures.")
+                if len(unique_tests) > _MAX_UNIQUE_TESTS:
+                    fail_msgs.append(
+                        f"{indent_by(1)}* ... and "
+                        f"{len(unique_tests) - _MAX_UNIQUE_TESTS} more tests.")
```
Comment (sriyerg) [inline util/dvsim/SimCfg.py:602]: I see you kept it simple by not counting the absolute number of remaining failing tests in the bucket but just the number of the remaining unique ones. That LGTM.

**Pull request merged**